Migrate Baw Client Routes

Allcharles opened the pull request:
# Migration of Baw Client Routes

Create all of the routing required to handle displaying the baw client pages. This makes use of the `BawClientComponent` created in the parent PR.

## Changes

- Added the following pages:
  - Listen
    - Listen Page (`/listen`) (#206)
    - Play Page (`/listen/:audioRecordingId`) (#207)
  - Library
    - Library Page (`/library`) (#208)
    - Annotation Page (`/library/:audioRecordingId/audio_events/:audioEventId`) (#209)
  - Visualize Page (`/visualize`) (#210)
  - Audio Analysis
    - List Page (`/audio_analysis`) (#211)
    - New Page (`/audio_analysis/new`) (#212)
    - Details Page (`/audio_analysis/:analysisJobId`) (#213)
    - Results Page (`/audio_analysis/:analysisJobId/results`) (#214)

## Problems

- Some menu routes are still wrong (#772)
- Identified component required redesign to simplify (#778 )
- Have not migrated citizen science pages, this will be a separate PR
- Test migrated pages load similar to https://github.com/QutEcoacoustics/workbench-client/blob/f6cb05d26af94923200c88d213c231d5053dbeef/src/app/components/shared/baw-client/baw-client.component.spec.ts#L133-L142
- ~~Still haven't dealt with #779~~ Closed in #807
- Potential issue with library link on secondary menu. When accessing an annotation inside the library, the library menu item will update to include the audio recording id in the query parameters. This means, if the user clicks back to the library, it will be prefilled and search for the audio recording. Depending on perspective, this is a bug or a feature
- ~~Add step in CI to download and build baw-client. This will allow for unit tests on whether the migration is working in the current build~~ Closed in #839 

## Visual Changes

### Listen Page (`/listen`)

![image](https://user-images.githubusercontent.com/3955116/107321537-72759000-6aee-11eb-80ab-2a165e53ecfa.png)

### Play Page (`/listen/:audioRecordingId`)

![image](https://user-images.githubusercontent.com/3955116/107321573-815c4280-6aee-11eb-8ab2-53c04ce78966.png)

### Library Page (`/library`)

![image](https://user-images.githubusercontent.com/3955116/107321596-8d480480-6aee-11eb-870f-c33b1aca1285.png)

### Annotation Page (`/library/:audioRecordingId/audio_events/:audioEventId`)

![image](https://user-images.githubusercontent.com/3955116/107321634-9c2eb700-6aee-11eb-9ada-9efed5727b65.png)

### Visualize Page (`/visualize`)

![image](https://user-images.githubusercontent.com/3955116/107321797-eb74e780-6aee-11eb-9c43-f8fe7dc0c285.png)

### Audio Analysis

#### List Page (`/audio_analysis`)

![image](https://user-images.githubusercontent.com/3955116/107321834-ffb8e480-6aee-11eb-81b5-ee7f31ff5408.png)

#### New Page (`/audio_analysis/new`)

![image](https://user-images.githubusercontent.com/3955116/107321855-0ba4a680-6aef-11eb-8bb6-067f7e84c1ec.png)

#### Details Page (`/audio_analysis/:analysisJobId`)

![image](https://user-images.githubusercontent.com/3955116/107321878-17906880-6aef-11eb-9e13-00c06976a3a6.png)

#### Results Page (`/audio_analysis/:analysisJobId/results`)

![image](https://user-images.githubusercontent.com/3955116/107321914-2840de80-6aef-11eb-8919-95cc36fdc3e0.png)

## Final Checklist

- [x] Assign reviewers if you have permission
- [x] Assign labels if you have permission
- [x] Link issues related to PR
- [x] Ensure project linter is not producing any warnings (`npm run lint`)
- [x] Ensure build is passing on all browsers (`npm run test:all`)
- [ ] Ensure CI build is passing
- [ ] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

Allcharles committed:
Build Baw Client (#839)

Build baw client in github actions, this allows the unit tests to validate that the baw client has been correctly integrated into the project.

Allcharles committed:
Migrate Baw Client Routes (#780)

Create all of the routing required to handle displaying the baw client pages. This makes use of the `BawClientComponent` created in #727.

atruskie commented on src/app/components/projects/site-card/site-card.component.ts:
```diff
@@ -41,10 +41,21 @@ import { Site } from "@models/Site";
           </li>
           <li *ngIf="site" class="nav-item">
-            <a id="play" class="nav-link" [routerLink]="site.playUrl">
+            <!-- Play link if recording exists -->
+            <a
+              *ngIf="recording"
+              id="play"
+              class="nav-link"
+              [routerLink]="recording?.viewUrl"
+            >
```
Should fallback to "No audio yet" if no audio link available. Maybe with a disabled play icon still?

atruskie commented on src/app/components/sites/site/site.component.html:
```diff
@@ -71,33 +86,35 @@ <h3>Recent Annotations</h3>
-        </li>
+              <!-- Annotation details -->
+              <ng-template #loaded>
+                "{{ tagging.tag?.text }}" by
+                {{ tagging.tag?.creator?.userName }}
```
these should ideally be both links to the library list page that pre-populate the search filter. Too hard?

Allcharles replied:
The current implementation on the ecosounds website is to link to the listen page, not the library

atruskie replied:
ah i missed the anchor in wrapper for the template. disregard

atruskie commented on src/app/components/sites/site/site.component.ts:
```diff
@@ -27,10 +27,10 @@ class SiteComponent extends PageComponent implements OnInit {
+  public newestRecording: AudioRecording;
   public marker: List<MapMarkerOption>;
-  public taggings: Tagging[];
+  public events: AudioEvent[];
```
For clarity: suggest name is `recentAudioEvents`

atruskie commented on src/app/components/sites/site/site.component.ts:
```diff
@@ -62,12 +68,19 @@ class SiteComponent extends PageComponent implements OnInit {
+            // Don't insert events with no tags
+            if (event.taggings.length === 0) {
+              continue;
+            }
```
Why?

atruskie commented on src/app/components/visualize/visualize.menus.ts:
```diff
@@ -0,0 +1,36 @@
+    }
+  }
+);
+
```
Missing QSP `siteIds : Array<Id>`

While flat parameters are the literal correct choice, this might feel better if implemented as a discriminated union for the parameters `siteId`, `siteIds`, `projectIds`, and `regionId` because they are exclusive choices.

Additionally, `regionId` is not currently supported by the old client... so perhaps throw error if we try to form such a link? Or better yet rewrite usages of `regionId` into `siteIds` QSP.

atruskie commented on src/app/models/AudioEvent.ts:
```diff
@@ -79,11 +80,18 @@ export class AudioEvent extends AbstractModel implements IAudioEvent {
 
   public get listenViewUrl(): string {
-    return listenMenuItem.uri([]);
+    // TODO Add start and end query params
```
blocked by issue # ?

atruskie commented on src/app/models/Bookmark.ts:
```diff
@@ -63,12 +63,14 @@ export class Bookmark extends AbstractModel implements IBookmark {
     console.warn("Bookmark viewUrl not implemented.");
-    return listenMenuItem.uri();
+    // TODO This link is wrong
+    return listenMenuItem.route.toRouterLink();
```
these are also blocked by the QSP  issue?

atruskie commented on src/app/services/baw-api/audio-recording/audio-recordings.service.ts:
```diff
@@ -53,6 +54,31 @@ export class AudioRecordingsService extends ReadonlyApi<AudioRecording> {
+    filters: Filters<IAudioRecording>,
+    region: IdOr<Region>
+  ) {
+    if (typeof region === "number") {
```
Should be using a type guard here https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards

atruskie commented on src/app/services/baw-api/baw-api.service.ts:
```diff
@@ -281,24 +281,62 @@ export abstract class BawApiService<Model extends AbstractModel> {
+   * @param models Foreign key values (if undefined, returns base filters)
+   * @param comparison Comparison to be performed
+   */
+  protected filterByForeignKeys(
```
I don't really like the foreign key term in this case. It is very database specific.

It's more like "filter by association"  right?

`filterThroughAssociation`?

atruskie commented on src/app/services/baw-api/baw-api.service.ts:
```diff
@@ -281,24 +281,62 @@ export abstract class BawApiService<Model extends AbstractModel> {
+   */
+  protected filterByForeignKeys(
+    filters: Filters<Model>,
+    key: KeysOfType<Model, number | string | Set<string> | Set<number>>,
```
This type definition is complex. 

I'd suggest pulling it out into a type alias, or simplifying to a duck-type (via an interface)

atruskie reviewed (changes requested):
I'd say a lot of these non-CRUD pages don't fit well into the list/details nomenclature we've used for CRUD pages. Listen, visualise are more composite pages that tie more than one basic resource together. Not important enough to change now, might be worth considering later.

There are several places with TODOs for malformed URLs (good). Just add issue link to each so they're findable using the same pattern.

Great work, minor changes suggested

atruskie reviewed (approved):
💯

Allcharles committed:
Working Login (#853)

* Cleanup app styling to create full height components

- Cleanup CSS styling for the app component so that the default router outlet component will have access to the full height of the page
- Created basic audio analysis page

* Working example

Couple bugs to fix
- AngularJS links are not reloading the parent page
- IFrame does not resize to content which causes the scrollbar to be in the wrong location

* Created new audio analysis page

* Cleanup iframe module

- Disabled broken tests
- Renamed IFrameComponent to BawClientComponent
- Isolated BawClientComponent into a module
- Created placeholder unit tests
- Updated interceptor to include cookies in requests

* Implemented BawClientComponent unit tests

* Fixed broken tests and rebase conflicts

* Fixed broken unit tests

Discovered bug in how secondary menu unit tests were saving/restoring the defaultMenu global variable (converted Lists to arrays, but not back)

* Fix site/region details tag links

* Working login request

* Update security service tests

* Cleanup interceptor tests

* Fix rebase issues

* Cleanup api request responseType's

This should make things a lot more clear for when the interceptor should set accept/content-type headers to requests

* Cleanup review comments
